File: app/src/main/java/com/example/david/lists/view/common/TouchHelperCallback.java
package com.example.david.lists.view.common;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.RecyclerView;

public final class TouchHelperCallback extends ItemTouchHelper.Callback {


    public interface MovementCallback {
        void dragging(int fromPosition, int toPosition);

        void movedPermanently(int newPosition);
    }


    private final MovementCallback movementCallback;

    public TouchHelperCallback(MovementCallback movementCallback) {
        this.movementCallback = movementCallback;
    }


    @Override
    public int getMovementFlags(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder) {
        return makeMovementFlags(
                ItemTouchHelper.UP | ItemTouchHelper.DOWN,
                0
        );
    }

    @Override
    public boolean isLongPressDragEnabled() {
        return false;
    }

    @Override
    public boolean isItemViewSwipeEnabled() {
        return false;
    }


    @Override
    public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
        movementCallback.dragging(viewHolder.getAdapterPosition(), target.getAdapterPosition());
        return true;
    }

    @Override
    public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
    }


    /**
     * This is called post {@link TouchHelperCallback#onMove(RecyclerView, RecyclerView.ViewHolder, RecyclerView.ViewHolder)}
     * and {@link TouchHelperCallback#onSwiped(RecyclerView.ViewHolder, int)}.
     */
    @Override
    public void clearView(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder) {
        super.clearView(recyclerView, viewHolder);
        movementCallback.movedPermanently(viewHolder.getAdapterPosition());
    }
}

File: app/src/main/java/com/example/david/lists/view/common/TouchHelperCallback.kt
package com.example.david.lists.view.common

import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView

class TouchHelperCallback(private val movementCallback: MovementCallback) : ItemTouchHelper.Callback() {

    interface MovementCallback {
        fun dragging(fromPosition: Int, toPosition: Int)

        fun movedPermanently(newPosition: Int)
    }


    override fun getMovementFlags(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) =
            makeMovementFlags(
                    ItemTouchHelper.UP or ItemTouchHelper.DOWN,
                    0
            )

    override fun isLongPressDragEnabled() = false

    override fun isItemViewSwipeEnabled() = true


    override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
        movementCallback.dragging(viewHolder.adapterPosition, target.adapterPosition)
        return true
    }

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {}


    /**
     * This is called post [TouchHelperCallback.onMove] and [TouchHelperCallback.onSwiped].
     */
    override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
        super.clearView(recyclerView, viewHolder)
        movementCallback.movedPermanently(viewHolder.adapterPosition)
    }
}
